test/unit-tests/expression/parse.test.js

```diff
@@ -2605,4 +2615,228 @@ describe('parse', function () {
+      const outerCond = math.parse('true ? (false ? 1 : 2) : 3')
+      const outerSources = [
+        { index: 5, text: '?' },
+        { index: 23, text: ':' }
+      ]
+      assert.deepStrictEqual(outerCond.sources, outerSources)
+
+      const innerCond = outerCond.trueExpr.content
+      const innerSources = [
+        { index: 14, text: '?' },
+        { index: 18, text: ':' }
+      ]
+      assert.deepStrictEqual(innerCond.sources, innerSources)
+    })
+
+    it('adds sources for assignments', function () {
+      const parsed = math.parse('val = 42')
+      assert.deepStrictEqual(parsed.sources, [{ index: 4, text: '=' }])
+    })
+
+    it('adds sources for percents', function () {
+      const parsed = math.parse('13%')
+      assert.deepStrictEqual(parsed.sources, [{ index: 2, text: '%' }])
+    })
+
+    it('adds sources for implicit multiplication', function () {
+      const parsed = math.parse('2a')
+
+      // index is where the multiplication symbol would be
+      assert.deepStrictEqual(parsed.sources, [{ index: 1, text: '' }])
+    })
+
+    it('adds sources for conversions', function () {
+      const parsedTo = math.parse('1 foot to in')
+      assert.deepStrictEqual(parsedTo.sources, [{ index: 7, text: 'to' }])
+
+      const parsedIn = math.parse('in in 1 foot')
+      assert.deepStrictEqual(parsedIn.sources, [{ index: 3, text: 'in' }])
+    })
+
+    it('adds sources for unary operators', function () {
+      const unaryPlus = math.parse('+1')
+      const unaryMinus = math.parse('-1')
+      assert.deepStrictEqual(unaryPlus.sources, [{ index: 0, text: '+' }])
+      assert.deepStrictEqual(unaryMinus.sources, [{ index: 0, text: '-' }])
+    })
+
+    it('adds sources for power operators', function () {
+      const parsed = math.parse('2^4')
+      assert.deepStrictEqual(parsed.sources, [{ index: 1, text: '^' }])
+    })
+
+    it('adds sources for constants', function () {
+      const parsedTrue = math.parse('true')
+      assert.deepStrictEqual(parsedTrue.sources, [{ index: 0, text: 'true' }])
+
+      const parsedNull = math.parse('null')
+      assert.deepStrictEqual(parsedNull.sources, [{ index: 0, text: 'null' }])
+
+      const parsedInfinity = math.parse('Infinity')
+      assert.deepStrictEqual(parsedInfinity.sources, [{ index: 0, text: 'Infinity' }])
+
+      const parsedNaN = math.parse('NaN')
+      assert.deepStrictEqual(parsedNaN.sources, [{ index: 0, text: 'NaN' }])
+    })
+
+    it('adds sources for function calls', function () {
+      const parsed = math.parse('foo(1, 2)')
+
+      // should have sources for parens and each param delimeter
+      const sources = [
+        { index: 3, text: '(' },
+        { index: 5, text: ',' },
+        { index: 8, text: ')' }
+      ]
+      assert.deepStrictEqual(parsed.sources, sources)
+    })
+
+    it('adds sources for string literals', function () {
+      const singleQuote = math.parse("'hello'")
```

Comment:
I think in case of a string, the string value itself should also be an entry in the sources, right?